staging/vhost-device-gpu/src/lib.rs

```diff
@@ -42,58 +42,260 @@ pub mod protocol;
+            gpu_mode,
+            capset_mask: capset,
+            flags,
+        })
     }
 
-    /// Return the path of the unix domain socket which is listening to
-    /// requests from the guest.
-    pub fn socket_path(&self) -> &Path {
-        &self.socket_path
+    pub fn requires_render_server(&self) -> bool {
+        self.gpu_mode == GpuMode::VirglRenderer && self.capset_mask.contains(GpuCapset::VENUS)
     }
 
     pub const fn gpu_mode(&self) -> GpuMode {
         self.gpu_mode
     }
+
+    pub const fn capsets(&self) -> GpuCapset {
+        self.capset_mask
+    }
+
+    pub const fn flags(&self) -> &GpuFlags {
+        &self.flags
+    }
 }
 
 #[cfg(test)]
 mod tests {
-    use tempfile::tempdir;
+    use assert_matches::assert_matches;
 
     use super::*;
 
     #[test]
-    fn test_gpu_config() {
-        // Test the creation of `GpuConfig` struct
-        let test_dir = tempdir().expect("Could not create a temp test directory.");
-        let socket_path = test_dir.path().join("socket");
-        let gpu_config = GpuConfig::new(socket_path.clone(), GpuMode::VirglRenderer);
-        assert_eq!(gpu_config.socket_path(), socket_path);
+    fn test_gpu_config_create_default_virglrenderer() {
+        let config = GpuConfig::new(GpuMode::VirglRenderer, None, GpuFlags::new_default()).unwrap();
+        assert_eq!(config.gpu_mode, GpuMode::VirglRenderer);
+        assert_eq!(config.capsets(), GpuConfig::DEFAULT_VIRGLRENDER_CAPSET_MASK);
+    }
+
+    #[test]
+    fn test_gpu_config_create_default_gfxstream() {
+        let config = GpuConfig::new(GpuMode::Gfxstream, None, GpuFlags::default()).unwrap();
+        assert_eq!(config.gpu_mode, GpuMode::Gfxstream);
+        assert_eq!(config.capsets(), GpuConfig::DEFAULT_GFXSTREAM_CAPSET_MASK);
+    }
+
+    #[test]
+    fn test_gpu_config_wrong_combination() {
+        let config = GpuConfig::new(GpuMode::VirglRenderer, None, GpuFlags::default()).unwrap();
+        assert_eq!(config.gpu_mode, GpuMode::VirglRenderer);
+    }
+
+    #[test]
+    fn test_gpu_config_wrong_combination_virglrenderer() {
+        let capset = GpuCapset::VIRGL2 | GpuCapset::GFXSTREAM_VULKAN;
+        let result = GpuConfig::new(
+            GpuMode::VirglRenderer,
+            Some(capset),
+            GpuFlags::new_default(),
+        );
+        assert_matches!(
+            result,
+            Err(GpuConfigError::CapsetUnsuportedByMode(
+                GpuMode::VirglRenderer,
+                GpuCapset::GFXSTREAM_VULKAN
+            ))
+        );
+    }
+
+    #[test]
+    fn test_gpu_config_wrong_combination_gfxstream() {
+        let capset = GpuCapset::VIRGL2 | GpuCapset::GFXSTREAM_VULKAN;
+        let result = GpuConfig::new(GpuMode::Gfxstream, Some(capset), GpuFlags::new_default());
+        assert_matches!(
+            result,
+            Err(GpuConfigError::CapsetUnsuportedByMode(
+                GpuMode::Gfxstream,
+                GpuCapset::VIRGL2
+            ))
+        );
+    }
```

Comment:
maybe we should reduce redundancy: 
```suggestion 

    fn assert_invalid_gpu_config(mode: GpuMode, capset: GpuCapset, expected_capset: GpuCapset) {
        let result = GpuConfig::new(mode, Some(capset), GpuFlags::new_default());
        assert_matches!(
            result,
            Err(GpuConfigError::CapsetUnsuportedByMode(
                mode,
                unsupported_capset
            )) if unsupported_capset == expected_capset
        );
    }
    
    #[test]
    fn test_gpu_config_valid_combination() {
        let config = GpuConfig::new(GpuMode::VirglRenderer, None, GpuFlags::default()).unwrap();
        assert_eq!(config.gpu_mode, GpuMode::VirglRenderer);
    }
    
    #[test]
    fn test_gpu_config_invalid_combinations() {
        assert_invalid_gpu_config(
            GpuMode::VirglRenderer,
            GpuCapset::VIRGL2 | GpuCapset::GFXSTREAM_VULKAN,
            GpuCapset::GFXSTREAM_VULKAN,
        );
    
        assert_invalid_gpu_config(
            GpuMode::Gfxstream,
            GpuCapset::VIRGL2 | GpuCapset::GFXSTREAM_VULKAN,
            GpuCapset::VIRGL2,
        );
    }

Reply:
I also noticed it increased the test coverage